include/sharpSAT/structures.h

```diff
@@ -273,8 +273,36 @@ class ClauseHeader {
     creation_time_ = time;
   }
 
+  /*!
+   * Number of entries in `std::vector<LiteralID>` to fit a \ref ClauseHeader.
+   * 
+   * This version checks necessary bounds and if not, throws `std::logic_error`.
+   * Since this is `noexcept`, runtime calls `std::terminate` immediately.
+   * [Trick source](http://ericniebler.com/2014/09/27/assert-and-constexpr-in-cxx11/)
+   */
   constexpr static unsigned overheadInLits() {
-    return sizeof(ClauseHeader) / sizeof(LiteralID);
+    return (overheadImpl() == 0 || overheadImpl() * sizeof(LiteralID) < sizeof(ClauseHeader))
+         ? throw assert_failure("ClauseHeader::overheadInLits is mal-implemented.")
+         : overheadImpl();
+  }
+
+private:
```

Comment:
You should be able to skip all this just by using `static_assert` for the check, as `sizeof(T)` is always\* a compile time constant.

\* VLAIS do not count as they are only implemented by gcc.

Reply:
You mean this?
```C++
static_assert(overheadImpl() == 0 || overheadImpl() * sizeof(LiteralID) < sizeof(ClauseHeader));
return overheadImpl();
```

This gives me the `non-constant condition for static assertion` error. :frowning_face:

Reply:
If you use Clang, it also tells you
```
/mnt/c/ubuntu/sharpSAT/include/sharpSAT/structures.h:285:21: note: undefined function 'overheadImpl' cannot be used in a constant expression
/mnt/c/ubuntu/sharpSAT/include/sharpSAT/structures.h:304:29: note: declared here
  constexpr static unsigned overheadImpl() noexcept {
                            ^
```
The fix is easy: move `overheadImpl()` ahead of the function that uses it. (I would also move the static assert into separate TU from the header and have the function itself without any checking).

Reply:
[Nailed it.](https://img.buzzfeed.com/buzzfeed-static/static/enhanced/webdr03/2013/1/25/14/enhanced-buzz-13312-1359142735-3.jpg?downsize=715:*&output-format=auto&output-quality=auto) :sunglasses: 

```c++
namespace overhead {

  //! Number of entries in `std::vector<U>` to fit a `T`.
  template<class T, class U>
  constexpr unsigned calculate() noexcept {
    return (sizeof(T) + sizeof(U) - 1) / sizeof(U);
  }

  /*!
   * Number of entries in `std::vector<U>` to fit a `T`.
   * 
   * This version checks soundness.
   */
  template<
    class T, class U,
    class = typename std::enable_if< (calculate<T,U>() != 0)
      && (calculate<T,U>() * sizeof(U) >= sizeof(T)) >::type>
  constexpr unsigned calculate_and_validate() noexcept {
    return calculate<T,U>();
  }
}; // overhead
```